### PR TITLE
Mortar variables setup

### PR DESCRIPTION
**Pull request by dewenyushu**
- Issue description:
In the. add_variables task the contact action requires the displacement variables to be already set up to infer the order of the Mortar variables. However moose can't guarantee the action execution order within a given moose task. #14630 

- Development:
Created a new task, add_mortar_variable, in the mortar action to ensure that the Lagrange multipliers are created after the displacement variables being setup

### CODE REVIEW

**Comment by permcody** [comment]
I have sent you an invite to join Moose users so your PRs will auto test. Please accept the invite.

File: framework/src/base/Moose.C [hunk 1]
```diff
@@ -270,6 +272,7 @@ addActionTypes(Syntax & syntax)
                            "(init_displaced_problem)"
                            "(add_aux_variable, add_variable, add_elemental_field_variable,"
                            " add_external_aux_variables)"
+                           "add_mortar_variable"
```
Comment (dschwen) [inline framework/src/base/Moose.C]: I think this needs parenthesis...

File: modules/contact/src/actions/ContactAction.C [hunk 2]
```diff
@@ -210,6 +213,23 @@ ContactAction::addMortarContact()
...
+    // determine necessary order
+    const bool second = _problem->mesh().hasSecondOrderElements();
+
+    params.set<MooseEnum>("order") = second ? "SECOND" : "FIRST";
+    params.set<MooseEnum>("family") = "LAGRANGE";
+
+    // Loop through the displacement variables
+    for (const auto & disp : displacements)
+    {
+      // Create displacement variables
```
Comment (dschwen) [inline modules/contact/src/actions/ContactAction.C]: The displacement variables are created by the TM master action, no need to duplicate that functionality in the contact action

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/src/base/Moose.C (hunk 1 above)
Comment (dewenyushu) [reply framework/src/base/Moose.C]: Agree

File: modules/contact/src/actions/ContactAction.C (hunk 2 above)
Comment (dewenyushu) [reply modules/contact/src/actions/ContactAction.C]: Makes sense 👍 This is now resolved

**Comment by dschwen** [comment]
Ok, squash your three commits into one and add a test, then this should be good to go.

File: framework/src/base/Moose.C [hunk 3]
```diff
@@ -111,6 +111,7 @@ addActionTypes(Syntax & syntax)
   registerMooseObjectTask("add_variable",                 MooseVariableBase,      false);
   registerMooseObjectTask("add_aux_variable",             MooseVariableBase,      false);
   registerMooseObjectTask("add_elemental_field_variable", MooseVariableBase,      false);
+  registerMooseObjectTask("add_mortar_variable",          MooseVariableBase,      false);
```
Comment (dschwen) [inline framework/src/base/Moose.C]: I don't think this line should be here (it only applies to stuff in the `[Variables]` block if I'm not mistaken).

File: modules/combined/test/tests/contact/horizontal_blocks_mortar_TM.i [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+[GlobalParams]
+  displacements = 'disp_x disp_y'
+  volumetric_locking_correction = true
+[]
+
+[Mesh]
+  [./original_file_mesh]
+    type = FileMeshGenerator
+    file = horizontal_blocks.e
```
Comment (dschwen) [inline modules/combined/test/tests/contact/horizontal_blocks_mortar_TM.i]: This could easily be done using mesh modifiers to save some repo space.

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/src/base/Moose.C (hunk 3 above)
Comment (dewenyushu) [reply framework/src/base/Moose.C]: That is correct. Thanks for pointing this out!

File: modules/combined/test/tests/contact/horizontal_blocks_mortar_TM.i (hunk 4 above)
Comment (dewenyushu) [reply modules/combined/test/tests/contact/horizontal_blocks_mortar_TM.i]: Makes sense. I have switched to a generated mesh and deleted the original mesh file.

**Review by dschwen — APPROVED** [review]

**Pull request merged**